### PR TITLE
Add statistics option for stage durations

### PR DESCRIPTION
**Pull request by lzap**
This adds a new option that prints statistics. There is a small refactoring and enhancements of the progress bar in several separate commits. I took inspiration from https://github.com/osbuild/image-builder-cli/pull/189/commits/e142ee344153daacd0474a045a57c623c8eb9bfe but that commit was missing tests so I did not pulled that in and did this myself.

The first commit upgrades to the latest `images` library, I was not sure what to do with the repommd option so setting that to `nil`. This probably needs a followup.

Then a new option is added that collects data from osbuild (recent version needed). The goal is to print the top 10 stages at the end of the build and here is a problem I need help with. No matter how I approach this, the output seems to be always broken. There must be some mutex or goroutine hanging I do not see:

```
OSBUILD_COMMAND=/home/lzap/bin/osbuild ibcli build --distro fedora-41 --with-statistics minimal-raw
...
[\] Image building step
[7 / 7] Pipeline xz [--------------------------------------------------------------------------------------------------------------------------------------------->] 100.00%
Statistics:ge org.osbuild.xz [------------------------------------------------------------------------------------------------------------------------------------>] 100.00%
- Finished module org.osbuild.rpm: 1m9.76825761s
- Finished module org.osbuild.xz: 24.227047158s
- Finished module org.osbuild.rpm: 9.729729055s
- Finished module org.osbuild.selinux: 6.208979558s
- Finished module org.osbuild.copy: 3.116474187s
[|] Image building step
[7 / 7] Pipeline xz [--------------------------------------------------------------------------------------------------------------------------------------------->] 100.00%
[2 / 2] Stage org.osbuild.xz [------------------------------------------------------------------------------------------------------------------------------------>] 100.00%
Message: Finished pipeline xz
```

The current approach is to add a new method to the progress interface that allows to schedule output after `Stop()` is called. It does not work for some reason. I could refactor the methods to return results in an argument or via return value as well that could do it. But this is where I need a suggestion which direction should I take.

### CODE REVIEW

File: pkg/progress/command.go [hunk 1]
```diff
@@ -108,6 +110,13 @@ func runOSBuildNoProgress(pb ProgressBar, manifest []byte, exports []string, opt
 
 var osbuildCmd = "osbuild"
 
+func init() {
+	// allow overriding for development purposes (not useful for end users)
+	if v := os.Getenv("OSBUILD_COMMAND"); v != "" {
```
[the remark this answers is not in the document]
Comment (lzap) [reply pkg/progress/command.go:115]: Not sure how a mock is related, I want to use a different `osbuild` binary with some patches applied locally.

[the remark this answers is not in the document]
Comment (supakeen) [reply pkg/progress/command.go:115]: This can be done through normal `PATH` adjustments right? E.g. `PATH=/somepath/containing/osbuild:$PATH image-builder ...`. That's how I usually do it at least.

**Comment by lzap** [comment]
Oh also squashed the `go.mod` change for `x/term` because images were already upgraded to v0.109 in the meantime.

**Comment by lzap** [comment]
Filed one additional change for `images` which adds pipeline name for better experience: https://github.com/osbuild/images/pull/1885

I will take a look on the test, I would prefer to keep it in one PR this is not a feature that needs to be rushed.

**Comment by lzap** [comment]
I have found the index of bounds issue, it was actually trivial. If the terminal is not present, width was reported as zero. But the `ShortenString` utility function had a bug, I fixed it and when the width is zero it returns the very same string without any shortening. Added more unit tests to cover the edge cases including zero width, one character width etc.

This probably means the test that is trying out the progress bar is executed in an environment that does not report proper terminal size, which is no surprise for CICD. Quick googling reveals that we should consider adding `--terminal-size 80x24` to the `podman` command for those tests that show progress bars. I think the test might not even testing anything on CICD at all, not sure what is the behavior for the progress library we use in case of non-interactive terminals.

**Comment by supakeen** [comment]
> I have found the index of bounds issue, it was actually trivial. If the terminal is not present, width was reported as zero. But the `ShortenString` utility function had a bug, I fixed it and when the width is zero it returns the very same string without any shortening. Added more unit tests to cover the edge cases including zero width, one character width etc.
> 
> This probably means the test that is trying out the progress bar is executed in an environment that does not report proper terminal size, which is no surprise for CICD. Quick googling reveals that we should consider adding `--terminal-size 80x24` to the `podman` command for those tests that show progress bars. I think the test might not even testing anything on CICD at all, not sure what is the behavior for the progress library we use in case of non-interactive terminals.

Things like these (and there are likely more lurking, terminals are very finicky) is why I'd really like to not have the terminal width detection changes in this PR but a separate one where it can be fully considered.

The *rest* of the PR looks great.

**Comment by supakeen** [comment]
> I fully agree and as much as I would like to split the second commit in this series does a big split of `progress.go` and `command.go` and then the last commit edits both. I mean this must be unresolvable via git and the only option would be to fully rewrite this.

You are doing three separate things in a single PR. A splitting of files, a change in how we handle terminal output, and then the actual change you and I want to land: statistics output.

This is clearly not ideal and has happened in other PRs before as well. I'm really not happy to shoehorn in extra changes I'm not on board with just to get in the one change I do agree with.

If it's too much work for you to detangle this I can do that work for you. Let me know and I'll force-push over this branch and open a separate PR for the terminal changes.
 
> I do not know how can I do the mentioned `Schutzfile` change, I do not see such file in this repo. The current patch needs v0.109 but I guess next week there will be another release already with the pipeline name so I can rebase this one.

Yea sorry, this repository doesn't use that. My bad.

File: pkg/progress/progress.go [hunk 2]
```diff
@@ -144,21 +146,20 @@ func (b *terminalProgressBar) SetProgress(subLevel int, msg string, done int, to
...
 
-func shorten(msg string) string {
-	msg = strings.Replace(msg, "\n", " ", -1)
-	// XXX: make this smarter
-	if len(msg) > 60 {
-		return msg[:60] + "..."
+func getTerminalWidth() int {
+	width, _, err := term.GetSize(int(os.Stdout.Fd()))
+	if err != nil {
+		width = 80
```
[the remark this answers is not in the document]
Comment (mvo5) [reply pkg/progress/progress.go:161]: Maybe something like this:
```go
diff --git a/pkg/progress/progress.go b/pkg/progress/progress.go
index 564cc10..3194cd2 100644
--- a/pkg/progress/progress.go
+++ b/pkg/progress/progress.go
@@ -164,22 +164,22 @@ func getTerminalWidth() int {
 }
 
 func (b *terminalProgressBar) SetPulseMsgf(msg string, args ...any) {
-       b.spinnerPb.Set("spinnerMsg", util.ShortenString(fmt.Sprintf(msg, args...), getTerminalWidth()-10))
+       b.spinnerPb.Set("spinnerMsg", fmt.Sprintf(msg, args...))
 }
 
 func (b *terminalProgressBar) SetMessagef(msg string, args ...any) {
-       b.msgPb.Set("msg", util.ShortenString(fmt.Sprintf(msg, args...), getTerminalWidth()-10))
+       b.msgPb.Set("msg", fmt.Sprintf(msg, args...))
 }
 
 func (b *terminalProgressBar) render() {
        var renderedLines int
-       fmt.Fprintf(b.out, "%s%s\n", ERASE_LINE, b.spinnerPb.String())
+       fmt.Fprintf(b.out, "%s%s\n", ERASE_LINE, util.ShortenString(b.spinnerPb.String(), getTerminalWidth()))
        renderedLines++
        for _, prog := range b.subLevelPbs {
                fmt.Fprintf(b.out, "%s%s\n", ERASE_LINE, prog.String())
                renderedLines++
        }
-       fmt.Fprintf(b.out, "%s%s\n", ERASE_LINE, b.msgPb.String())
+       fmt.Fprintf(b.out, "%s%s\n", ERASE_LINE, util.ShortenString(b.msgPb.String(), getTerminalWidth()))
        renderedLines++
        fmt.Fprint(b.out, cursorUp(renderedLines))
 }
``` 
ie. shorten after the render.
